typescript/deploy-sdk/src/ism/generic-ism.ts

```diff
@@ -0,0 +1,173 @@
+ *
+ * @example
+ * ```typescript
+ * const reader = createIsmReader(chainMetadata, chainLookup);
+ * const ismConfig = await reader.read(ismAddress);
+ * ```
+ */
+export function createIsmReader(
+  chainMetadata: ChainMetadataForAltVM,
+  chainLookup: ChainLookup,
+): IsmReader {
+  const protocolProvider = getProtocolProvider(chainMetadata.protocol);
+  const artifactManager: IRawIsmArtifactManager =
+    protocolProvider.createIsmArtifactManager(chainMetadata);
+
+  return new IsmReader(artifactManager, chainLookup);
+}
+
+/**
+ * Converts a DeployedIsmArtifact to DerivedIsmConfig format.
+ * This handles the conversion between the new Artifact API and the old Config API.
+ */
+function artifactToDerivedConfig(
+  artifact: DeployedIsmArtifact,
+): DerivedIsmConfig {
+  const config = artifact.config;
+  const address = artifact.deployed.address;
+
+  // For routing ISMs, recursively convert nested artifacts
+  if (config.type === AltVM.IsmType.ROUTING) {
+    const domains: DomainRoutingIsmConfig['domains'] = {};
+    for (const [domainId, nestedArtifact] of Object.entries(config.domains)) {
+      if (isArtifactUnderived(nestedArtifact)) {
+        domains[domainId] = nestedArtifact.deployed.address;
+      } else if (isArtifactDeployed(nestedArtifact)) {
+        domains[domainId] = artifactToDerivedConfig(nestedArtifact);
+      }
+      // Note: ArtifactState.NEW should never occur in expanded routing configs
+    }
+    return {
+      type: 'domainRoutingIsm',
+      owner: config.owner,
+      domains,
+      address,
+    } satisfies Extract<DerivedIsmConfig, DomainRoutingIsmConfig>;
+  }
+
+  // For other ISM types, just add the address
+  return {
+    ...config,
+    address,
+  };
+}
+
+/**
+ * Generic ISM Reader that can read any ISM type by detecting its type
+ * and recursively expanding nested ISMs (e.g., for routing ISMs).
+ */
+export class IsmReader
+  implements ArtifactReader<IsmArtifactConfig, DeployedIsmAddresses>
+{
+  protected readonly logger: Logger = rootLogger.child({
+    module: IsmReader.name,
+  });
+
+  constructor(
+    protected readonly artifactManager: IRawIsmArtifactManager,
+    protected readonly chainLookup: ChainLookup,
+  ) {}
+
+  async read(address: string): Promise<DeployedIsmArtifact> {
+    // Read once via readIsm() - detects type
+    const { artifactState, config, deployed } =
+      await this.artifactManager.readIsm(address);
+
+    // For routing ISMs, expand nested domain ISMs recursively
+    if (config.type === AltVM.IsmType.ROUTING) {
+      return this.expandRoutingIsm({ artifactState, config, deployed });
+    }
+
+    // For non-routing ISMs, the raw and expanded configs are identical
+    return {
+      artifactState,
+      config,
+      deployed,
+    };
+  }
+
+  /**
+   * Expands a raw routing ISM config by recursively reading the domain ISMs.
+   * Takes a pre-read raw artifact to avoid double reading.
+   */
+  private async expandRoutingIsm(
```

Comment:
Moved this here because with this PR having a compound RoutingIsmReader doesn't work well and seems unnecessary. The read() method wasn't type-safe and the whole class wasn't used. This remains stitched in RoutingIsmWriter but it may also go away as a class. Besides, we considered removing inheritance between ArtifactWriter and Reader.